### PR TITLE
Add BERT Base to Composer

### PR DESCRIPTION
**Pull request by moinnadeem**
# Motivation
Composer currently contains autoregressive language models, but masked language models are just as important for speeding up training. This PR serves to add the first of masked language models (BERT) to Composer.

# Merging Criterion
- [x] Demonstrate that BERT achieves acceptable downstream GLUE performance on all benchmarks
- [ ] Demonstrate average GLUE performance on validation set and test set
- [x] Code quality should be up-to-par, with well-commented abstractions
- [ ] New TorchMetrics that have been added should include tests
- [x] New YAML files should be clean and not contain any unnecessary hparams

### CODE REVIEW

**Comment by moinnadeem** [comment]
As a heads up, I'm ripping out the `duration` abstraction that I created. It's useful for internal experimentation, but I think Ravi's Time abstraction will be a superset of its features. If anyone thinks that it should be included, feel free to mention it and I can include it

**Comment by moinnadeem** [comment]
Cool, this is ready for review!

**To the reviewers:** Let me know if you want any WandB notebooks on hyperparameters that I have swept for pre-training. I ended up sweeping learning rate and Adam's beta2, so we have WandB experiments for that. Additionally, we have WandB experiments for all of the GLUE downstream tasks. It might be good to verify that I copied the right hyperparameters from there.

I ended up leaving the ``duration`` abstraction in, and fixed the unit tests for that. This will be ripped out soon in favor of Ravi's time abstraction, but it could be useful until then.

Let me know if you have any questions! The final performance is available in the #r-nlp channel in Slack if anyone wants to view it.

File: composer/yamls/models/glue/mnli-m.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+train_dataset:
```
Comment (abhi-mosaic) [inline composer/yamls/models/glue/mnli-m.yaml:1]: What's the purpose of this `mnli-m` YAML config? Seems like a duplicate.

**Review by abhi-mosaic — COMMENTED** [review]
Couple comments here and there, most are MosaicTransformer / BERT refactoring things that can be addressed in a future PR. Some more important questions re. loss functions:

 1) The GLUE datasets use truncating-and-padding, is the `BERTForSequenceClassification` model equipped to ignore padding in its training loss calculations? I think this is probably ok, as `torch.nn.CrossEntropyLoss` has a hardcoded default of `ignore_index: -100`.

2) Some of the training metrics being recorded are logically weird, like averaging train loss over the whole training run. We could instead add an `eval_train` Evaluator once we have support, which I think is probably what we want.

3) `CrossEntropyLoss` metric seems like it has a (currently) silent bug.

I didn't review much of the Time or Scheduler -related codepaths as I figure those will be being changed relatively soon anyways, and I trust @moinnadeem's testing as he has been actually launching the runs and monitoring.

File: composer/yamls/models/glue/cola.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+validate_every_n_batches: 250
+validate_every_n_epochs: 1
+callbacks:
+  - lr_monitor: {}
+save_checkpoint:
+  folder: glue_checkpoints
+  interval: 1
+  interval_unit: ep
+load_checkpoint:
+  filepath: /mnt/cota/moin/bert_checkpoint/bert_checkpoints/ep7.pt
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/yamls/models/glue/cola.yaml:75]: Some YAMLs also have / don't have this `load_checkpoint` field

File: composer/yamls/models/glue/mnli-m.yaml (hunk 1 above)
Comment (moinnadeem) [reply composer/yamls/models/glue/mnli-m.yaml:1]: Ah, it's two different datasets -- mnli matched versus mnli-mismatched. We can merge them when the "evaluate on multiple datasets" PR is merged in!

**Review by Landanjs — COMMENTED** [review]
Most of my comments are cosmetic things that can be addressed as you see fit. Couple of points:

- I think the NLP and CV CrossEntropyLoss can be generalized to a single metric, but an issue can be made for this to be done later.
- As Abhi mentioned, more clear code around the different BERT task types would be nice (i.e. select metric based on string names instead of number of labels).
- I did not thoroughly go through the scheduler code since it seems fairly extensive. Might be good for the engineering team to do a pass through to see if it can be simplified.
- Should we keep our data directories in the yaml files? If not, replace with an arbitrary directory or set to null?

File: composer/models/bert/bert_hparams.py [hunk 3]
```diff
@@ -74,7 +74,7 @@ def initialize_object(self) -> "MosaicTransformer":
         else:
             raise ValueError('One of pretrained_model_name or model_config needed.')
 
-        # set the number of labels ot the voacb size, used for measuring MLM accuracy
+        # set the number of labels to the voacb size, used for measuring MLM accuracy
```
Comment (Landanjs) [inline composer/models/bert/bert_hparams.py:77]: Also typo with vocab

File: composer/datasets/glue.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+            first_half = inp[text_column_names[0]]
+            second_half = inp[text_column_names[1]] if text_column_names[1] in inp else None
+            return self.tokenizer(
+                text=first_half,
+                text_pair=second_half,
+                padding="max_length",
+                max_length=self.max_seq_length,
+                truncation=True,
+            )
```
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/datasets/glue.py:75]: To be clear:
Each sample has a pair (A, B). Let's assume (A, B) are both tokenized.
A, B are concatenated with some special tokens, the format becomes `[CLS] (A tokens) [SEP] (B tokens) [SEP]`. Three additional tokens are added during this process -- the CLS token, and the two SEP tokens.

If `len(a_tokens) + len(b_tokens) + 3 > max_seq_len`, then the sequence is truncated.  If the sequence needs to be truncated, then tokens are iteratively removed from the longest sequence of the two, until `max_seq_len` is reached.

If `len(a_tokens) + len(b_tokens) + 3 < max_seq_len`, then `[PAD]` tokens are iteratively added after the end of the sentence until `max_seq_len` is reached.

The attention mask masks out the `[PAD]` tokens from the attention calculation such that the sequence does not attend to these items, and the loss is averaged only on the tokens for which the attention mask attends. Therefore, these padding tokens do not contribute to the loss. 

Does this make sense?

File: composer/models/bert/model.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        # we're going to remove the label from the expected inputs
+        # since we will handle metric calculation with TorchMetrics instead of HuggingFace.
+        self.model_inputs.remove("labels")
+
+        self.train_metrics = []
+        self.val_metrics = []
+
+        # TODO (Moin): make sure this is moved to be dataset-specific
+        # if config.num_labels=1, then we are training a regression task, so we should update our loss functions
```
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/models/bert/model.py:37]: > I think there should probably be something like config.task_type: ["regression", "binary", "classification", "mlm"] rather than just a check on config.num_labels. Then different workloads can specify in the BERTConfig what type of model head+loss they want to use.

It's important to note that the current task-specific metric calculation is just a temporary stopgap until the "evaluation on multiple datasets" PR is merged. At that point, the metrics and losses will be a function of the dataset, not the model. This will naturally give us what you are asking for, and I agree with your statement. I just implemented something quickly so when the multiple datasets PR is ready, we can switch over.

cc @Landanjs I think you had the same comment

File: composer/yamls/models/bert-base.yaml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    use_pretrained: false
+    tokenizer_name: bert-base-uncased
+    pretrained_model_name: bert-base-uncased
+optimizer:
+  decoupled_adamw:
+    lr: 5.0e-4
+    betas:
+      - 0.9
+      - 0.98
+    eps: 1.0e-06
```
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/yamls/models/bert-base.yaml:42]: I think you remember from the last BERT meeting that we have significant work to do on hyperparameter robustification, but we're going to punt to sweeping the cardinal product of all of these hparams to mid-February so we can use our A100 cluster!

Is it okay if we punt this until later? I can sweep it now, but we'll have to sweep it again later when we are evaluating multiple hyperparameters together.

File: composer/yamls/models/glue/cola.yaml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    shuffle: false
+    drop_last: false
+model:
+  bert_classification:
+    num_labels: 2
+    use_pretrained: true
+    tokenizer_name: bert-base-uncased
+    pretrained_model_name: bert-base-uncased
+optimizer:
+  adamw:
```
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/yamls/models/glue/cola.yaml:24]: Changed!

**Review by moinnadeem — COMMENTED** [review]
Overall, great feedback from the two of you! I think I've addressed every comment.

A few common things:
1) I agree that the implementation of BERT's multiple evaluation metrics is not complete. The reason is that it's a temporary implementation, and will be thrown out once the "evaluation on multiple datasets" PR is merged. Then, we'll shift all of this evaluation to the dataset, and we can avoid using hacks like _how many labels are in the task?_

2) Great attention to detail on docstrings and variable names @Landanjs , thanks for that!

Replying to comments:

## Landan's Comments
> I think the NLP and CV CrossEntropyLoss can be generalized to a single metric, but an issue can be made for this to be done later.

Yes, I'd agree there. I think it's also best once we have an implementation that we want it to work with, since I wouldn't know what CV would need.

> As Abhi mentioned, more clear code around the different BERT task types would be nice (i.e. select metric based on string names instead of number of labels).
See point #1! This is temporary, since it'll be moved to dataset-specific tasks soon.

> Should we keep our data directories in the yaml files? If not, replace with an arbitrary directory or set to null?
I have replaced it with an arbitrary directory

## Abhi's comments
> The GLUE datasets use truncating-and-padding, is the BERTForSequenceClassification model equipped to ignore padding in its training loss calculations? I think this is probably ok, as torch.nn.CrossEntropyLoss has a hardcoded default of ignore_index: -100.

Yes, it is setup to ignore the padding token! See the comment I made about -100 in your specific comment

> Some of the training metrics being recorded are logically weird, like averaging train loss over the whole training run. We could instead add an eval_train Evaluator once we have support, which I think is probably what we want.

See the comment about averaging training loss -- we actually aren't applying this to training code.

> CrossEntropyLoss metric seems like it has a (currently) silent bug.
Yes, I agree that it was a silent bug. It is resolved!

File: composer/yamls/models/glue/cola.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/yamls/models/glue/cola.yaml:75]: Two changes:
1) After discussion with the team, we changed this to be a publicly-accessible HTTPS url! Composer will support this
2) I've added this to every YAML

File: composer/models/bert/model.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        assert self.training is False, "For validation, model must be in eval mode"
+
+        # temporary hack until eval on multiple datasets is finished
+        labels = batch.pop('labels')
+        output = self.forward(batch)
+        output = output['logits']
+
+        # if we are in the single class case, then remove the classes dimension
+        if output.shape[1] == 1:
+            output = output.squeeze()
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply composer/models/bert/model.py:103]: @moinnadeem not sure if you missed this, but do you have thoughts here?

File: composer/datasets/glue.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/datasets/glue.py:75]: Yep sounds good to me!

File: composer/models/bert/model.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            self.val_acc = Accuracy()
+
+            self.train_matthews = MatthewsCorrcoef(num_classes=config.num_labels)
+            self.val_matthews = MatthewsCorrcoef(num_classes=config.num_labels)
+
+            self.train_metrics.extend([self.train_acc, self.train_matthews])
+            self.val_metrics.extend([self.val_acc, self.val_matthews])
+
+        if config.num_labels == len(self.tokenizer):  # tests for MLM pre-training
+            ignore_index = -100
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/models/bert/model.py:66]: Ack on the -100 convention being PyTorch convention. And yes a future refactor can move all the loss function code, don't need to address rn if it's working.

File: composer/yamls/models/bert-base.yaml [hunk 10]
```diff
@@ -0,0 +1,84 @@
+train_dataset:
+  lm:
+    split: train
+    datadir:
+      - /mnt/gcp/datasets/wikipedia_saved_bert_128
+      - /mnt/gcp/datasets/bookcorpus_saved_bert_128
+    tokenizer_name: bert-base-uncased
+    seed: 19
+    shuffle: false
+    drop_last: false
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/yamls/models/bert-base.yaml:10]: Hmm but there is an optimization issue with `drop_last` because the very last gradient step will be out-of-proportion. Consider if the usual batch size is 2048, but the very last incomplete batch is only like 12 or so. Then the last gradient will be many times larger than usual (becaus |g| goes down as BS goes up), and so the very last update could be much bigger than we expect. Even with an optimizer like Adam, this could still produce problems. Feels bad if the last step of training is potentially harmful :( 

For multi-epoch runs with `drop_last`, as we do in ImageNet, you still have shuffling each epoch, so in expectation all samples do get seen.

File: composer/models/bert/model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/models/bert/model.py:37]: Ack that the future Evaluator framework solves the metrics and loss functions. And I guess we won't have switches for heads within `BERTModel`, but instead multiple higher level classes like `BERTForClassification`?

File: composer/models/nlp_metrics.py [hunk 11]
```diff
@@ -3,12 +3,159 @@
...
+        """Updates the internal state with results from a new batch.
+
+        Args:
+            output (Mapping): The output from the model, which must contain
+                either the Tensor or a Mapping type that contains the loss or model logits.
+            target (Tensor): A Tensor of ground-truth values to compare against.
+        """
+        assert isinstance(self.predictions, list)
+        assert isinstance(self.labels, list)
+        self.predictions.append(output)
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply composer/models/nlp_metrics.py:143]: TBH, I think it would cleaner and fairly easy to calculate the true positive (tp), false positive (fp), and false negative (fn) rates in the update, then manually compute the F1 score tp / (tp + 0.5 * (fp + fn)) from [here](https://en.wikipedia.org/wiki/F-score)

```suggestion
        prediction = output.argmax(dim=1)
        self.tp += prediction[target].sum()
        self.fp += (prediction[target] == 0).sum()
        self.fn += (prediction[target == 0] == 1).sum()
```

Up to your discretion since this is already working 🙂

File: composer/yamls/models/bert-base.yaml [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  lm:
+    split: validation
+    datadir:
+      - /mnt/gcp/datasets/wikipedia_saved_bert_128
+      - /mnt/gcp/datasets/bookcorpus_saved_bert_128
+    tokenizer_name: bert-base-uncased
+    seed: 19
+    shuffle: false
+    drop_last: false 
+    subsample_ratio: 0.02
```
[the remark this answers is not in the document]
Comment (Landanjs) [reply composer/yamls/models/bert-base.yaml:26]: Thanks!

File: composer/yamls/models/glue/cola.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Landanjs) [reply composer/yamls/models/glue/cola.yaml:75]: Awesome! 👏

File: composer/models/bert/model.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+            self.train_loss = CrossEntropyLoss(ignore_index=ignore_index, vocab_size=config.num_labels)
```
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/models/bert/model.py:67]: OK gotcha... so I guess there's some naming conventions to clean up because `training_metrics` is really like `batch_metrics` and the usual metrics are more like dataset `eval_metrics`... and then again batch metrics don't really need to be metrics at all... making my head spin but should be assigned in a separate issue :P

File: composer/yamls/models/bert-base.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (abhi-mosaic) [reply composer/yamls/models/bert-base.yaml:42]: Yep punt for later sounds good! Just caught my eye and wanted to highlight.

File: composer/models/nlp_metrics.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/models/nlp_metrics.py:143]: The more we talk about metrics, the more I realize that we should fix it. I'll implement your suggestion + add unit tests against sklearn's implementation today!

**Review by abhi-mosaic — APPROVED** [review]
Left 2 converations open just for considerations but nothing blocking. LGTM!

File: composer/models/bert/model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/models/bert/model.py]: @Landanjs I just added this change!

File: composer/models/nlp_metrics.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/models/nlp_metrics.py:138]: @Landanjs I just adopted your suggestion and ensured that it properly works!

File: composer/models/nlp_metrics.py [hunk 14]
```diff
@@ -3,12 +3,159 @@
...
+        self.labels.append(target)
+
+    def compute(self) -> float:
+        """Aggregate the state over all processes to compute the metric.
+
+        Returns:
+            loss (Tensor): The loss averaged across all batches.
+        """
+        # take the argmax to get label indicies
+        predictions = torch.Tensor(self.predictions)
```
[the remark this answers is not in the document]
Comment (moinnadeem) [reply composer/models/nlp_metrics.py]: @Landanjs Heads up that `torch.Tensor(...)` has been removed!

File: composer/models/nlp_metrics.py [hunk 15]
```diff
@@ -9,6 +9,149 @@
...
+        """Aggregate the state over all processes to compute the metric.
+
+        Returns:
+            loss (Tensor): The loss averaged across all batches.
+        """
+        # Return average loss over entire dataset
+        return self.sum_loss / self.total_items  #type: ignore (third-party)
+
+
+# TODO (Moin): write tests for this!
```
Comment (Landanjs) [inline composer/models/nlp_metrics.py:107]: Ah last thing (I think), could you remove these TODOs?

Comment (moinnadeem) [reply composer/models/nlp_metrics.py:107]: ooh good catch! Let me actually grep for TODOs overall

Comment (moinnadeem) [reply composer/models/nlp_metrics.py]: Done!

**Comment by Landanjs** [comment]
Mostly reviewed new changes to metrics. LGTM!

**Review by Landanjs — APPROVED** [review]
